akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/PersistentActorImpl.scala

```diff
@@ -143,8 +143,15 @@ import akka.{ actor ⇒ a }
   }
 
   def applySideEffect(effect: ChainableEffect[_, S]): Unit = effect match {
-    case _: Stop.type @unchecked ⇒ context.stop(self)
-    case SideEffect(callbacks)   ⇒ callbacks.apply(state)
+    case _: Stop.type @unchecked ⇒
+      context.stop(self)
+    case Unstash(commands) ⇒
+      // Process unstashed commands before taking new commands from the mailbox.
+      // TODO: this does not correctly deal with the situation that people unstash while
+      // processing a stashed command, but let's write a test before building that.
+      commands.foreach(receiveCommand)
```

Comment:
Hehe yeah... that dance is a bit involved... I hope to complete my initial work today so this could be implemented in the new internals